### PR TITLE
refactor: Refactoring Round 2

### PR DESCRIPTION
**Pull request by xavierchanth**
**- What I did**

- Separated `lib/` into the following:
  - `sshnp/`
  - `sshnpd/`
  - `sshrvd/`
  - `common/`
  - `version.dart` -> only file which sits in the root of `lib/`

- Extracted out Arg parsing and at_client creation from SSHNP and SSHNPD classes
  - N.B. sshnp's is a little bit more complicated due to --config-files, which is a meta parameter
  - at_client creation has some weird parameters and could probably be cleaner, but it get's the job done, it can be further refined later
- Extracted interface of SSHNP and SSHNPD (spec/impl files now)
- FIX: fixed a bug where the atSigns would be left unset in sshnpd and sshrvd if the key file path was specified

- Extracted sshrvd to a spec / impl / params class similar to sshnpd

**- How I did it**

**- How to verify it**

**- Description for the changelog**
refactor: Refactoring Round 2

### CODE REVIEW

**Comment by xavierchanth** [comment]
It's always the simple things, like not assigning a variable, which break everything...

**Review by gkc — DISMISSED** [review]
Nice work Xavier, LGTM. Thank you for the many individual commits, makes the reviewer's job so much easier

**Comment by xavierchanth** [comment]
> Nice work Xavier, LGTM. Thank you for the many individual commits, makes the reviewer's job so much easier

Of course, it also makes everyone's lives easier if we uncover a bug or something and need to look back through change history.

File: packages/sshnoports/bin/sshnp.dart
Comment (JeremyTubongbanua) [inline packages/sshnoports/bin/sshnp.dart:24]: Weren't there issues with this exit(0)?

**Review by JeremyTubongbanua — CHANGES_REQUESTED** [review]
If possible, could we merge https://github.com/atsign-foundation/sshnoports/pull/258 first? This would be able to do a small test on SSHRVD which was refactored in this PR

File: packages/sshnoports/bin/sshnp.dart
Comment (xavierchanth) [reply packages/sshnoports/bin/sshnp.dart:24]: No, sshnp needs to exit

Comment (xavierchanth) [reply packages/sshnoports/bin/sshnp.dart:24]: It will still keep the background isolate running which is what we want, but if it doesn't exit, then it won't exit automatically until that background isolate has died on its own.

**Comment by xavierchanth** [comment]
> If possible, could we merge #258 first? This would be able to do a small test on SSHRVD which was refactored in this PR

Yes, let's do it

**Comment by xavierchanth** [comment]
wait for the tests on trunk to pass before merging

**Comment by xavierchanth** [comment]
@JeremyTubongbanua need your review to unblock the merge

**Review by JeremyTubongbanua — DISMISSED** [review]

**Comment by JeremyTubongbanua** [comment]
@XavierChanth Did you pull latest trunk?

**Comment by xavierchanth** [comment]
> @XavierChanth Did you pull latest trunk?

Thanks for the reminder, will do it now

**Comment by xavierchanth** [comment]
sshrvd is failing because the option was --keyFile and I changed it to --key-file for consistency... what do we want to do about it?

**Comment by xavierchanth** [comment]
> sshrvd is failing because the option was --keyFile and I changed it to --key-file for consistency... what do we want to do about it?

cc: @gkc @cconstab @JeremyTubongbanua 

Ah, I could add "keyFile" as an alias for "key-file" to keep things backward compatible.

**Comment by JeremyTubongbanua** [comment]
Sorry @XavierChanth I'm confused why this is failing. I'm not using the `-k` flag at all

This is what the container is running
```sh
~/.local/bin/sshrvd -a @8485wealthy51 -i $(hostname -i) -v -s
```

This is the error
```
sshrvd  | Starting OpenBSD Secure Shell server: sshd
sshrvd  | .
sshrvd  | Invalid argument(s): Could not find an option named "keyFile".
sshrvd  | Version : 3.4.0
sshrvd  | -k, --key-file              atSign's atKeys file if not in ~/.atsign/keys/
sshrvd  | -a, --atsign (mandatory)    atSign for sshrvd
sshrvd  | -m, --manager               Managers atSign that sshrvd will accept requests from. Default is any atSign can use sshrvd
sshrvd  |                             (defaults to "open")
sshrvd  | -i, --ip (mandatory)        FQDN/IP address sent to clients
sshrvd  | -v, --[no-]verbose          More logging
sshrvd  | -s, --[no-]snoop            Snoop on traffic passing through service
sshrvd  | Unhandled exception:
sshrvd  | Invalid argument(s): Could not find an option named "keyFile".
sshrvd  | #0      ArgResults.[] (package:args/src/arg_results.dart:65)
sshrvd  | #1      new SSHRVDParams.fromArgs (package:sshnoports/sshrvd/sshrvd_params.dart:30)
sshrvd  | #2      SSHRVDImpl.fromCommandLineArgs (package:sshnoports/sshrvd/sshrvd_impl.dart:53)
sshrvd  | #3      SSHRVD.fromCommandLineArgs (package:sshnoports/sshrvd/sshrvd.dart:[44](https://github.com/atsign-foundation/sshnoports/actions/runs/5577702284/jobs/10190828045?pr=257#step:6:45))
sshrvd  | #4      main (file:///app/repo/bin/sshrvd.dart:5)
sshrvd  | #5      _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:294)
sshrvd  | #6      _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:189)
```

**Comment by xavierchanth** [comment]
I saw the error, let me check that it's not my code given the command you've provided

**Comment by xavierchanth** [comment]
@JeremyTubongbanua it was my code:

```dart
deviceAtsign = r['atsign'];
    managerAtsign = r['manager'];
    atKeysFilePath =
        r['keyFile'] ?? getDefaultAtKeysFilePath(homeDirectory, deviceAtsign);
```

Missed a spot... last line should be r['key-file']

**Comment by xavierchanth** [comment]
Oops, I did half the fix in one file, and half the fix in the other file. Both rvd and npd will have the "keyFile" alias

**Comment by xavierchanth** [comment]
Looks like I have some more work to do, I need to trace what is being sent between the isolates and figure out what is causing the issue here: https://github.com/atsign-foundation/sshnoports/actions/runs/5578079267/jobs/10191890499?pr=257

**Comment by xavierchanth** [comment]
I am very confused right now... it's saying that I'm trying to send a SecureSocket
(line 2 scroll across)
```
sshrvd  | Unhandled exception:
sshrvd  | Invalid argument(s): Illegal argument in isolate message: object is unsendable - Library:'dart:io' Class: _SecureSocket@14069316 (see restrictions listed at `SendPort.send()` documentation for more information)
sshrvd  |  <- Instance of 'OutboundConnectionImpl' (from package:at_lookup/src/connection/outbound_connection_impl.dart)
sshrvd  |  <- Instance of 'AtLookupImpl' (from package:at_lookup/src/at_lookup_impl.dart)
sshrvd  |  <- Instance of 'RemoteSecondary' (from package:at_client/src/client/remote_secondary.dart)
sshrvd  |  <- Instance of 'AtClientImpl' (from package:at_client/src/client/at_client_impl.dart)
sshrvd  |  <- Instance of 'SSHRVDImpl' (from package:sshnoports/sshrvd/sshrvd_impl.dart)
sshrvd  |  <- Context num_variables: 1
sshrvd  |  <- Closure: ((int, int, String, String, bool)) => void from Function '_socketConnector@19502927':. (from dart:core)
sshrvd  | 
sshrvd  | #0      Isolate._spawnFunction (dart:isolate-patch/isolate_patch.dart:392)
sshrvd  | #1      Isolate.spawn (dart:isolate-patch/isolate_patch.dart:372)
sshrvd  | #2      SSHRVDImpl._spawnSocketConnector (package:sshnoports/sshrvd/sshrvd_impl.dart:179)
sshrvd  | #3      SSHRVDImpl._notificationHandler (package:sshnoports/sshrvd/sshrvd_impl.dart:[133](https://github.com/atsign-foundation/sshnoports/actions/runs/5591810712/jobs/10223412252?pr=257#step:6:133))
sshrvd  | #4      _RootZone.runUnaryGuarded (dart:async/zone.dart:1594)
sshrvd  | #5      _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:339)
sshrvd  | #6      _DelayedData.perform (dart:async/stream_impl.dart:515)
sshrvd  | #7      _PendingEvents.handleNext (dart:async/stream_impl.dart:620)
sshrvd  | #8      _PendingEvents.schedule.<anonymous closure> (dart:async/stream_impl.dart:591)
sshrvd  | #9      _microtaskLoop (dart:async/schedule_microtask.dart:40)
sshrvd  | #10     _startMicrotaskLoop (dart:async/schedule_microtask.dart:49)
sshrvd  | #11     _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:123)
sshrvd  | #12     _Timer._runTimers (dart:isolate-patch/timer_impl.dart:405)
sshrvd  | #13     _Timer._handleMessage (dart:isolate-patch/timer_impl.dart:429)
sshrvd  | #14     _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:189)
```

**Comment by xavierchanth** [comment]
My mistake was that I made the function running on the separate isolate an instance function of the SSHNP class. It must either be a static function or top-level function. When it's an instance function there becomes some shared memory issues with the class that it's a part of.

**Comment by xavierchanth** [comment]
Ready for a final review @gkc @JeremyTubongbanua

**Review by JeremyTubongbanua — APPROVED** [review]

**Review by gkc — APPROVED** [review]
LGTM

**Pull request merged**
